### ISSUE TITLE
Character select often out of order upon first join

### ISSUE THREAD

**Issue by in1tiate** [AttorneyOnline/AO2-Client#122]
I'm not sure what sort of multithread hacking is going on in charselect.cpp, but it's causing character lists to be out of order when joining a server. I tested this on multiple servers but the result is consistently inconsistent. I've made a GIF because the issue can be hard to spot without side-by-side comparison:

[comparison GIF](https://i.imgur.com/SJ7pAEL.gif)

As seen in the GIF, the change can be radical or small, but the character select will be out of order until the user has selected a character and then hit the 'change character' button.

**Comment by stonedDiscord** [comment]
this seems related to Courtroom::filter_character_list()

**Comment by oldmud0** [comment]
Char select buttons are created on a per-thread basis. I frankly don't know why I allowed this code to go through.

The append is called inside the thread body: 
https://github.com/AttorneyOnline/AO2-Client/blob/4b0110d96cb3ad383d8ca517faa3c5235bed21df/src/charselect.cpp#L25

This means that not only does the order of characters become nondeterministic, but it is also ripe for crashes because there is no construct being used to prevent concurrent access. Use a [mutex](https://doc.qt.io/qt-5/qmutex.html), or better, add the character buttons to the list before running the thread.

**Comment by likeawindrammer** [comment]
> This means that not only does the order of characters become nondeterministic, but it is also ripe for **crashes**

[backtrace.txt](https://github.com/AttorneyOnline/AO2-Client/files/4464035/backtrace.txt)

**Comment by oldmud0** [comment]
```
#2  0x00007ffff627bf6c in qErrnoWarning(char const*, ...) () at /home/user/Qt/5.12.7/gcc_64/lib/libQt5Core.so.5
#3  0x00007ffff627b6b4 in (anonymous namespace)::Q_QGS_qMessagePattern::innerFunction()::HolderBase::~HolderBase() [clone .isra.27] () at /home/user/Qt/5.12.7/gcc_64/lib/libQt5Core.so.5
#4  0x000055555557ff0f in QVector<AOCharButton*>::at(int) const (this=0x555555e3bff8, i=1066) at ../../../Qt/5.12.7/gcc_64/include/QtCore/qvector.h:429
```
Curious, what did it try to print out in the warning? Looks like the QVector got corrupted.

**Comment by cents02** [comment]
I believe @likeawindrammer who is handling the QA can confirm that this bug no longer exists

**Comment by in1tiate** [comment]
Perhaps not this bug, but a related one has been happening to me for a while now: https://youtu.be/5f136r4SrGI
I'm running Qt 5.14.1 on this particular environment, which may be the cause.

**Comment by oldmud0** [comment]
Fixed in #137.
